File: dev/executor/supervisor.ts
import { mkdirSync, renameSync, unlinkSync, watch } from "node:fs";
import { ensureConfigFile, readConfig } from "./config";
import {
	configDir,
	configFile,
	configName,
	configWatchDebounceMs,
	runtimeStateFile,
	safetyPollIntervalMs,
	stateDir,
} from "./paths";
import type {
	ManagedProcessSnapshot,
	NormalizedConfig,
	NormalizedInstance,
} from "./types";
import { CliError } from "./utils";
import { ProcessManager, type ManagedProcess } from "./process-manager";
import { LocalProxy } from "./local-proxy";

interface ManagedEntry {
	name: string;
	specKey: string;
	process: ManagedProcess;
}

type ReconcileAction =
	| {
			kind: "start";
			name: string;
			instance: NormalizedInstance;
			specKey: string;
	  }
	| {
			kind: "replace";
			name: string;
			instance: NormalizedInstance;
			specKey: string;
			current: ManagedEntry;
	  }
	| {
			kind: "stop";
			name: string;
			current: ManagedEntry;
	  };

interface ConfigCache {
	fingerprint: string | null;
	config: NormalizedConfig | null;
	issue: string;
}

interface RuntimeStateFile {
	version: 1;
	supervisor: {
		pid: number;
		version: string;
		startedAt: number;
	};
	instances: Record<string, ManagedProcessSnapshot>;
	proxy: {
		pendingRequests: number;
		pendingWebSockets: number;
	};
}

function specKey(
	config: NormalizedConfig,
	instance: NormalizedInstance,
): string {
	return JSON.stringify({
		dir: instance.dir,
		cmd: instance.cmd,
		env: instance.env,
		restart: config.restartTokens.get(instance.name) ?? "",
	});
}

function buildReconcilePlan(
	managed: Map<string, ManagedEntry>,
	config: NormalizedConfig,
): ReconcileAction[] {
	const plan: ReconcileAction[] = [];
	const desired = new Set<string>();

	for (const [name, instance] of config.instances) {
		desired.add(name);
		const current = managed.get(name);
		if (!config.isEnabled(name)) {
			if (current) plan.push({ kind: "stop", name, current });
			continue;
		}

		const nextSpecKey = specKey(config, instance);
		if (!current) {
			plan.push({ kind: "start", name, instance, specKey: nextSpecKey });
		} else if (current.specKey !== nextSpecKey) {
			plan.push({
				kind: "replace",
				name,
				instance,
				specKey: nextSpecKey,
				current,
			});
		}
	}

	for (const [name, current] of managed) {
		if (!desired.has(name)) {
			plan.push({ kind: "stop", name, current });
		}
	}

	return plan;
}

function errorText(error: unknown): string {
	if (error instanceof Error) return error.message;
	return String(error);
}

function safeIssueText(error: unknown): string {
	return errorText(error).replace(/\s+/g, " ").trim().slice(0, 300);
}

function reportIssue(
	issues: Map<string, string>,
	key: string,
	message: string,
): void {
	if (issues.get(key) === message) return;
	issues.set(key, message);
	console.error(message);
}

async function sourceFingerprint(): Promise<string> {
	const configText = await Bun.file(configFile).text();
	const controlFile = `${configFile}\0${
		(await Bun.file(`${stateDir}/control.json`).exists())
			? await Bun.file(`${stateDir}/control.json`).text()
			: ""
	}`;
	return String(Bun.hash(`${configText}\0${controlFile}`));
}

async function loadConfig(cache: ConfigCache): Promise<{
	config: NormalizedConfig | null;
	issue: string;
	changed: boolean;
}> {
	let fingerprint: string;
	try {
		fingerprint = await sourceFingerprint();
	} catch (error) {
		const issue = `Unable to read executor config: ${safeIssueText(error)}`;
		return { config: cache.config, issue, changed: true };
	}

	if (cache.fingerprint === fingerprint) {
		return { config: cache.config, issue: cache.issue, changed: false };
	}
	cache.fingerprint = fingerprint;

	try {
		const config = await readConfig(false);
		if (!config) {
			cache.issue = `Missing config file: ${configFile}`;
			return { config: cache.config, issue: cache.issue, changed: true };
		}
		cache.config = config;
		cache.issue = "";
		return { config, issue: "", changed: true };
	} catch (error) {
		const issue =
			error instanceof CliError
				? error.message
				: `Invalid config file: ${safeIssueText(error)}`;
		cache.issue = issue;
		return { config: cache.config, issue, changed: true };
	}
}

async function reconcile(
	managed: Map<string, ManagedEntry>,
	pm: ProcessManager,
	proxy: LocalProxy,
	cache: ConfigCache,
	issues: Map<string, string>,
	shouldStop: () => boolean = () => false,
): Promise<void> {
	const loaded = await loadConfig(cache);
	if (loaded.issue) {
		reportIssue(issues, "config", loaded.issue);
		return;
	}
	issues.delete("config");
	if (!loaded.config || !loaded.changed) return;

	const config = loaded.config;
	proxy.update(config);
	const plan = buildReconcilePlan(managed, config);

	for (const action of plan) {
		if (shouldStop()) return;
		try {
			switch (action.kind) {
				case "start": {
					if (shouldStop()) return;
					const process = pm.start(action.instance);
					managed.set(action.name, {
						name: action.name,
						specKey: action.specKey,
						process,
					});
					break;
				}
				case "replace":
					await action.current.process.stop();
					managed.delete(action.name);
					if (shouldStop()) return;
					managed.set(action.name, {
						name: action.name,
						specKey: action.specKey,
						process: pm.start(action.instance),
					});
					break;
				case "stop":
					await action.current.process.stop();
					managed.delete(action.name);
					break;
			}
			issues.delete(`instance:${action.name}`);
		} catch (error) {
			reportIssue(
				issues,
				`instance:${action.name}`,
				`[executor] ${action.name}: ${safeIssueText(error)}`,
			);
		}
	}
}

async function writeRuntimeState(
	managed: Map<string, ManagedEntry>,
	proxy: LocalProxy,
	startedAt: number,
): Promise<void> {
	mkdirSync(stateDir, { recursive: true });
	const instances: Record<string, ManagedProcessSnapshot> = {};
	for (const [name, entry] of managed) {
		instances[name] = entry.process.snapshot;
	}
	const state: RuntimeStateFile = {
		version: 1,
		supervisor: {
			pid: process.pid,
			version: Bun.version,
			startedAt,
		},
		instances,
		proxy: proxy.snapshot,
	};
	const tmpFile = `${runtimeStateFile}.${process.pid}.${Date.now()}.tmp`;
	try {
		await Bun.write(tmpFile, `${JSON.stringify(state, null, 2)}\n`);
		renameSync(tmpFile, runtimeStateFile);
	} finally {
		try {
			unlinkSync(tmpFile);
		} catch {
			// The rename already removed the temporary name.
		}
	}
}

export async function runSupervisor(proxyPort = 80): Promise<void> {
	const env = Object.fromEntries(
		Object.entries(process.env).map(([key, value]) => [key, value ?? ""]),
	);
	await ensureConfigFile();

	const managed = new Map<string, ManagedEntry>();
	const startedAt = Date.now();
	const issues = new Map<string, string>();
	let proxy!: LocalProxy;
	let statusWriteTail: Promise<void> = Promise.resolve();
	const saveRuntimeState = (): void => {
		statusWriteTail = statusWriteTail
			.then(() => writeRuntimeState(managed, proxy, startedAt))
			.catch((error) => {
				reportIssue(
					issues,
					"runtime-state",
					`[executor] unable to write runtime state: ${safeIssueText(error)}`,
				);
			});
	};
	const pm = new ProcessManager(env, { onStateChange: saveRuntimeState });
	proxy = new LocalProxy(proxyPort, 5_000, (name) => {
		const restart = managed.get(name)?.process.restart();
		if (restart) {
			void restart.catch((error) =>
				reportIssue(
					issues,
					`instance:${name}`,
					`[executor] ${name}: ${safeIssueText(error)}`,
				),
			);
		}
	});

	const cache: ConfigCache = { fingerprint: null, config: null, issue: "" };
	let stopping = false;
	let dirty = true;
	let wakeResolver: (() => void) | null = null;
	let watcher: ReturnType<typeof watch> | null = null;
	let debounceTimer: ReturnType<typeof setTimeout> | null = null;

	const wake = (): void => {
		if (debounceTimer) {
			clearTimeout(debounceTimer);
			debounceTimer = null;
		}
		dirty = true;
		const resolve = wakeResolver;
		wakeResolver = null;
		resolve?.();
	};
	const scheduleWatcherWake = (): void => {
		if (debounceTimer) clearTimeout(debounceTimer);
		debounceTimer = setTimeout(() => {
			debounceTimer = null;
			wake();
		}, configWatchDebounceMs);
	};
	const onConfigEvent = (
		_event: string,
		filename: string | Buffer | null,
	): void => {
		if (filename && filename.toString() !== configName) return;
		scheduleWatcherWake();
	};
	try {
		watcher = watch(configDir, { persistent: false }, onConfigEvent);
		watcher.on("error", (error) => {
			if (watcher) {
				watcher.close();
				watcher = null;
			}
			reportIssue(
				issues,
				"watcher",
				`[executor] config watcher disabled: ${safeIssueText(error)}`,
			);
			wake();
		});
	} catch (error) {
		reportIssue(
			issues,
			"watcher",
			`[executor] config watcher unavailable: ${safeIssueText(error)}`,
		);
	}

	let stopPromise: Promise<void> | null = null;
	const stopAll = (): Promise<void> => {
		if (stopPromise) return stopPromise;
		stopping = true;
		wake();
		stopPromise = Promise.allSettled(
			[...managed.values()].map((entry) => entry.process.stop()),
		).then(() => {
			managed.clear();
			saveRuntimeState();
		});
		return stopPromise;
	};

	const onSighup = (): void => wake();
	const onSigint = (): void => {
		void stopAll();
	};
	const onSigterm = (): void => {
		void stopAll();
	};
	process.on("SIGHUP", onSighup);
	process.on("SIGINT", onSigint);
	process.on("SIGTERM", onSigterm);

	try {
		while (!stopping) {
			if (dirty) {
				dirty = false;
				try {
					await reconcile(
						managed,
						pm,
						proxy,
						cache,
						issues,
						() => stopping,
					);
				} catch (error) {
					reportIssue(
						issues,
						"reconcile",
						`[executor] reconcile failed: ${safeIssueText(error)}`,
					);
				}
				saveRuntimeState();
				if (dirty) continue;
			}

			const woke = await new Promise<boolean>((resolve) => {
				let finished = false;
				let timer: ReturnType<typeof setTimeout>;
				const finish = (byWake: boolean): void => {
					if (finished) return;
					finished = true;
					clearTimeout(timer);
					if (wakeResolver === onWake) wakeResolver = null;
					resolve(byWake);
				};
				const onWake = (): void => finish(true);
				timer = setTimeout(() => finish(false), safetyPollIntervalMs);
				wakeResolver = onWake;
			});
			if (!woke) dirty = true;
		}
	} finally {
		if (debounceTimer) clearTimeout(debounceTimer);
		watcher?.close();
		watcher = null;
		process.off("SIGHUP", onSighup);
		process.off("SIGINT", onSigint);
		process.off("SIGTERM", onSigterm);
		try {
			await proxy.stop();
		} finally {
			await stopAll();
			await statusWriteTail;
		}
	}
}

File: dev/executor/src/supervisor.rs
use crate::cli::CliError;
use crate::config::{ensure_config_file, read_config, source_fingerprint, NormalizedConfig, NormalizedInstance};
use crate::paths::{config_dir, config_name, runtime_state_file, SAFETY_POLL_INTERVAL_MS, CONFIG_WATCH_DEBOUNCE_MS};
use crate::pm::{ManagedProcess, ProcessManager, ProcessManagerOptions};
use crate::proxy::{LocalProxy, LocalProxyOptions};
use crate::runner::now_ms;
use crate::state::{write_runtime_state, RuntimeStateFile, SupervisorInfo};
use notify::{Config, RecommendedWatcher, RecursiveMode, Watcher};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::signal::unix::{signal, SignalKind};

struct ManagedEntry {
    spec_key: String,
    process: Arc<ManagedProcess>,
}

enum ReconcileAction {
    Start {
        name: String,
        instance: NormalizedInstance,
        spec_key: String,
    },
    Replace {
        name: String,
        instance: NormalizedInstance,
        spec_key: String,
    },
    Stop {
        name: String,
    },
}

fn spec_key(config: &NormalizedConfig, instance: &NormalizedInstance) -> String {
    format!(
        "{}|{}|{:?}|{}",
        instance.dir,
        instance.cmd,
        instance.env,
        config.restart_tokens.get(&instance.name).cloned().unwrap_or_default()
    )
}

fn build_reconcile_plan(
    managed: &HashMap<String, ManagedEntry>,
    config: &NormalizedConfig,
) -> Vec<ReconcileAction> {
    let mut plan = Vec::new();
    let mut desired = std::collections::HashSet::new();
    for (name, instance) in &config.instances {
        desired.insert(name.clone());
        let current = managed.get(name);
        if !config.is_enabled(name) {
            if current.is_some() {
                plan.push(ReconcileAction::Stop { name: name.clone() });
            }
            continue;
        }
        let next = spec_key(config, instance);
        match current {
            None => plan.push(ReconcileAction::Start {
                name: name.clone(),
                instance: instance.clone(),
                spec_key: next,
            }),
            Some(entry) if entry.spec_key != next => plan.push(ReconcileAction::Replace {
                name: name.clone(),
                instance: instance.clone(),
                spec_key: next,
            }),
            Some(_) => {}
        }
    }
    for name in managed.keys() {
        if !desired.contains(name) {
            plan.push(ReconcileAction::Stop { name: name.clone() });
        }
    }
    plan
}

pub async fn run_supervisor(proxy_port: u16) -> Result<(), CliError> {
    ensure_config_file().await?;
    let env: std::collections::BTreeMap<_, _> = std::env::vars().collect();
    let started_at = now_ms();
    let managed: Arc<Mutex<HashMap<String, ManagedEntry>>> = Arc::new(Mutex::new(HashMap::new()));
    let dirty = Arc::new(Mutex::new(true));
    let stopping = Arc::new(Mutex::new(false));
    let proxy = LocalProxy::bind(
        proxy_port,
        5_000,
        Some({
            let managed = managed.clone();
            Arc::new(move |name: String| {
                if let Some(entry) = managed.lock().unwrap().get(&name) {
                    let process = entry.process.clone();
                    tokio::spawn(async move {
                        let _ = process.restart().await;
                    });
                }
            })
        }),
        120_000,
        LocalProxyOptions::default(),
    )
    .await
    .map_err(|e| CliError::new(e.to_string()))?;

    let save = {
        let managed = managed.clone();
        let proxy_port_unused = proxy.port();
        let _ = proxy_port_unused;
        move |proxy: &LocalProxy| {
            let instances = managed
                .lock()
                .unwrap()
                .iter()
                .map(|(name, entry)| (name.clone(), entry.process.snapshot()))
                .collect();
            let state = RuntimeStateFile {
                version: 1,
                supervisor: SupervisorInfo {
                    pid: std::process::id(),
                    version: env!("CARGO_PKG_VERSION").into(),
                    started_at,
                },
                instances,
                proxy: proxy.snapshot(),
            };
            let _ = write_runtime_state(&state);
        }
    };

    let pm = ProcessManager::new(
        env,
        ProcessManagerOptions {
            on_state_change: Some(Arc::new({
                // runtime writes happen from reconcile loop
                || {}
            })),
            ..ProcessManagerOptions::default()
        },
    );

    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<()>();
    let tx2 = tx.clone();
    let mut watcher = RecommendedWatcher::new(
        move |_| {
            let _ = tx.send(());
        },
        Config::default(),
    )
    .ok();
    if let Some(watcher) = watcher.as_mut() {
        let _ = watcher.watch(&config_dir(), RecursiveMode::NonRecursive);
    }

    let mut sighup = signal(SignalKind::hangup()).map_err(|e| CliError::new(e.to_string()))?;
    let mut sigterm = signal(SignalKind::terminate()).map_err(|e| CliError::new(e.to_string()))?;
    let mut sigint = signal(SignalKind::interrupt()).map_err(|e| CliError::new(e.to_string()))?;
    let mut last_fp = String::new();
    let mut last_config: Option<NormalizedConfig> = None;

    loop {
        if *stopping.lock().unwrap() {
            break;
        }
        let is_dirty = *dirty.lock().unwrap();
        if is_dirty {
            *dirty.lock().unwrap() = false;
            match source_fingerprint() {
                Ok(fp) if fp == last_fp => {}
                Ok(fp) => {
                    last_fp = fp;
                    match read_config(false).await {
                        Ok(Some(config)) => {
                            last_config = Some(config.clone());
                            proxy.update(&config);
                            let plan = {
                                let guard = managed.lock().unwrap();
                                build_reconcile_plan(&guard, &config)
                            };
                            for action in plan {
                                match action {
                                    ReconcileAction::Start {
                                        name,
                                        instance,
                                        spec_key,
                                    } => {
                                        let process = pm.start(&instance);
                                        managed.lock().unwrap().insert(
                                            name,
                                            ManagedEntry { spec_key, process },
                                        );
                                    }
                                    ReconcileAction::Replace {
                                        name,
                                        instance,
                                        spec_key,
                                    } => {
                                        if let Some(current) = managed.lock().unwrap().remove(&name)
                                        {
                                            let _ = current.process.stop().await;
                                        }
                                        let process = pm.start(&instance);
                                        managed.lock().unwrap().insert(
                                            name,
                                            ManagedEntry { spec_key, process },
                                        );
                                    }
                                    ReconcileAction::Stop { name } => {
                                        if let Some(current) = managed.lock().unwrap().remove(&name)
                                        {
                                            let _ = current.process.stop().await;
                                        }
                                    }
                                }
                            }
                        }
                        Ok(None) => {}
                        Err(error) => eprintln!("{}", error.message),
                    }
                }
                Err(error) => eprintln!("{}", error.message),
            }
            save(&proxy);
        }

        tokio::select! {
            _ = sighup.recv() => { *dirty.lock().unwrap() = true; }
            _ = sigterm.recv() => { *stopping.lock().unwrap() = true; }
            _ = sigint.recv() => { *stopping.lock().unwrap() = true; }
            _ = rx.recv() => {
                tokio::time::sleep(Duration::from_millis(CONFIG_WATCH_DEBOUNCE_MS)).await;
                *dirty.lock().unwrap() = true;
            }
            _ = tokio::time::sleep(Duration::from_millis(SAFETY_POLL_INTERVAL_MS)) => {
                *dirty.lock().unwrap() = true;
            }
        }
        let _ = tx2;
        let _ = runtime_state_file();
        let _ = config_name();
        let _ = last_config.as_ref();
    }

    let _ = proxy.stop(None).await;
    let names: Vec<_> = managed.lock().unwrap().keys().cloned().collect();
    for name in names {
        if let Some(entry) = managed.lock().unwrap().remove(&name) {
            let _ = entry.process.stop().await;
        }
    }
    save(&proxy);
    Ok(())
}
